[Feature] Add index_size meta function to query column index sizes

zhan7236 opened the pull request:
## Why I'm doing:

This PR implements the feature requested in [https://github.com/StarRocks/starrocks/issues/62680](https://github.com/StarRocks/starrocks/issues/62680) to provide a meta function for querying column index sizes.

Users need a way to analyze and monitor the size of various indexes (BITMAP, BLOOM, ZONEMAP) on their table columns for storage optimization and troubleshooting purposes.

## What I'm doing:

Fixes [https://github.com/StarRocks/starrocks/issues/62680](https://github.com/StarRocks/starrocks/issues/62680). Add a new `index_size()` meta function that works with the `[_META_]` hint to query the size of column indexes.

### Syntax

```sql
SELECT index_size(column_ref [, index_type]) FROM table_name [_META_];
```

### Parameters

- `column_ref`: The column to check index size for
- `index_type` (optional): The type of index to query
  - `'BITMAP'`: Bitmap index size
  - `'BLOOM'`: Bloom filter index size
  - `'ZONEMAP'`: Zone map index size
  - `'ALL'` (default): Total size of all indexes

### Examples

```sql
-- Get total index size for column k1
SELECT index_size(k1) FROM my_table [_META_];

-- Get only ZONEMAP index size
SELECT index_size(k1, 'ZONEMAP') FROM my_table [_META_];

-- Get BLOOM filter index size
SELECT index_size(k2, 'BLOOM') FROM my_table [_META_];
```

## Changes Made

### Backend (C++)
- `be/src/storage/rowset/column_reader.h/cpp`: Added `get_index_size()` method to retrieve index sizes from metadata
- `be/src/storage/meta_reader.h/cpp`: Added `META_INDEX_SIZE` constant and `_collect_index_size()` function

### Frontend (Java)
- `fe/.../FunctionSet.java`: Registered `INDEX_SIZE` as builtin aggregate function
- `fe/.../PushDownAggToMetaScanRule.java`: Added support for index_size in meta scan push down
- `fe/.../RewriteSimpleAggToMetaScanRule.java`: Added support for index_size in simple agg rewrite

### Tests
- Added test cases in `test/sql/test_meta_scan/`

### Documentation
- Added function documentation in `docs/en/` and `docs/zh/`

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

N/A - New feature

## Checklist:

- [x] I have added test cases for my code
- [x] I have added documentation for my code

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces index_size(column[, 'BITMAP'|'BLOOM'|'ZONEMAP'|'ALL']) for META_SCAN, with planner pushdown/rewrite, backend collectors, docs, and tests.
> 
> - **Backend (BE)**:
>   - Add `META_INDEX_SIZE` and support in `SegmentMetaCollecter` (special parsing `index_size_<TYPE>_<COL>`, BIGINT output).
>   - Implement `_collect_index_size` (recursive) and `ColumnReader::get_index_size()` to sum BITMAP/BLOOM/ZONEMAP metadata sizes.
>   - Treat `index_size` like size counters when building result chunks.
> - **Frontend/Planner (FE)**:
>   - Register builtin `index_size` (1-arg total, 2-arg typed) in `FunctionSet`.
>   - Extend `PushDownAggToMetaScanRule` and `RewriteSimpleAggToMetaScanRule` to support `index_size`, encode meta column names (`index_size_<TYPE>_<COL>`), and rewrite to `SUM`.
> - **Docs**:
>   - Add EN/ZH docs for `index_size` usage and semantics.
> - **Tests**:
>   - Add META_SCAN SQL tests covering `index_size` for ALL/ZONEMAP/BLOOM/BITMAP.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4592f9c004351346a8260cbd6f1d83bc0140e341. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

alvin-celerdata commented:
@cursor review

wyb commented on docs/zh/sql-reference/sql-functions/utility-functions/index_size.md:
```diff
@@ -0,0 +1,57 @@
+displayed_sidebar: docs
+---
+
+# index_size
```
how about `column_index_size`?

zhan7236 committed:
[Feature] Add index_size meta function to query column index sizes

Add a new index_size() meta function that works with the [_META_] hint
to query the size of column indexes (BITMAP, BLOOM, ZONEMAP, ALL).

Changes:
- FE: Register INDEX_SIZE builtin function with two overloads
- FE: Update PushDownAggToMetaScanRule and RewriteSimpleAggToMetaScanRule
- BE: Add META_INDEX_SIZE and implement _collect_index_size in MetaReader
- BE: Implement ColumnReader::get_index_size() method
- Add SQL tests and documentation (EN/ZH)

Fixes https://github.com/StarRocks/starrocks/issues/62680

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

eshishki commented:
@zhan7236 are you still pursuing this? It's been idle ~4.7 months and is now CONFLICTING, so it needs a rebase. Also @wyb's suggestion to rename `index_size` → `column_index_size` is worth taking — it lines up with the existing per-column meta-functions (`column_size`, `column_compressed_size`). A rebase + rename would get this back in shape for review.
